### PR TITLE
Accelerate Vector128<long>::op_Multiply on x64

### PR DESCRIPTION
**Pull request by EgorBo**
This PR optimizes `Vector128` and `Vector256` multiplication for `long/ulong` when AVX512 is not presented in the system. It makes XxHash128 faster, see https://github.com/dotnet/runtime/pull/103555#issuecomment-2180622218

```cs
public Vector128<long> Foo(Vector128<long> a, Vector128<long> b) => a * b;
```

Current codegen on x64 cpu without AVX512:
```asm
; Method MyBench:Foo
       push     rsi
       push     rbx
       sub      rsp, 104
       mov      rbx, rdx
       mov      rdx, qword ptr [r8]
       mov      qword ptr [rsp+0x58], rdx
       mov      rdx, qword ptr [r9]
       mov      qword ptr [rsp+0x50], rdx
       mov      rdx, qword ptr [rsp+0x58]
       imul     rdx, qword ptr [rsp+0x50]
       mov      qword ptr [rsp+0x60], rdx
       mov      rsi, qword ptr [rsp+0x60]
       mov      rdx, qword ptr [r8+0x08]
       mov      qword ptr [rsp+0x40], rdx
       mov      rdx, qword ptr [r9+0x08]
       mov      qword ptr [rsp+0x38], rdx
       mov      rcx, qword ptr [rsp+0x40]
       mov      rdx, qword ptr [rsp+0x38]
       call     [System.Runtime.Intrinsics.Scalar`1[long]:Multiply(long,long):long]   ;;; not inlined call!
       mov      qword ptr [rsp+0x48], rax
       mov      rax, qword ptr [rsp+0x48]
       mov      qword ptr [rsp+0x20], rsi
       mov      qword ptr [rsp+0x28], rax
       vmovaps  xmm0, xmmword ptr [rsp+0x20]
       vmovups  xmmword ptr [rbx], xmm0
       mov      rax, rbx
       add      rsp, 104
       pop      rbx
       pop      rsi
       ret      
; Total bytes of code: 120
```

New codegen:
```asm
; Method MyBench:Foo
       vmovups  xmm0, xmmword ptr [r8]
       vmovups  xmm1, xmmword ptr [r9]
       vpmuludq xmm2, xmm1, xmm0
       vpshufd  xmm1, xmm1, -79
       vpmulld  xmm0, xmm1, xmm0
       vxorps   xmm1, xmm1, xmm1
       vphaddd  xmm0, xmm0, xmm1
       vpshufd  xmm0, xmm0, 115
       vpaddq   xmm0, xmm0, xmm2
       vmovups  xmmword ptr [rdx], xmm0
       mov      rax, rdx
       ret      
; Total bytes of code: 50
```

### CODE REVIEW

**Comment by EgorBo** [comment]
@EgorBot -arm64 -profiler
```cs
using System.IO.Hashing;
using BenchmarkDotNet.Attributes;

public class Bench
{
    static readonly byte[] Data = GetData();

    private static byte[] GetData()
    {
        byte[] bytes = new byte[1000000];
        bytes.AsSpan().Fill(42);
        return bytes;
    }

    [Benchmark]
    public byte[] BenchXxHash128()
    {
        XxHash128 hash = new();
        hash.Append(Data);
        return hash.GetHashAndReset();
    }
}
```

**Comment by EgorBo** [comment]
Note: results should be better if we do it in JIT, it will enable loop hoisting, cse, etc for MUL

**Comment by neon-sunset** [comment]
Note https://github.com/dotnet/runtime/issues/103539#issuecomment-2172562760 (and https://godbolt.org/z/eqsrf341M) from xxHash128 issue.

**Comment by EgorBo** [comment]
@EgorBot -arm64
```cs
using System.IO.Hashing;
using BenchmarkDotNet.Attributes;

public class Bench
{
    static readonly byte[] Data = GetData();

    private static byte[] GetData()
    {
        byte[] bytes = new byte[1000000];
        bytes.AsSpan().Fill(42);
        return bytes;
    }

    [Benchmark]
    public byte[] BenchXxHash128()
    {
        XxHash128 hash = new();
        hash.Append(Data);
        return hash.GetHashAndReset();
    }
}
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128_1.cs [hunk 1]
```diff
@@ -274,10 +272,19 @@ public static Vector128<T> Zero
...
                         a.GetLower().AsUInt32(), b.GetLower().AsUInt32()).AsUInt16(),
                     AdvSimd.MultiplyWideningUpper(
-                        a.AsUInt32(),
-                        b.AsUInt32()).AsUInt16()).AsUInt64().As<ulong, T>();
+                        a.AsUInt32(), b.AsUInt32()).AsUInt16()).AsUInt64().As<ulong, T>();
+            }
+            if (Sse41.IsSupported)
+            {
+                Vector128<ulong> a = left.AsUInt64();
+                Vector128<ulong> b = left.AsUInt64();
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128_1.cs:280]: Same issue for Arm64:
```suggestion
                Vector128<ulong> a = left.AsUInt64();
                Vector128<ulong> b = right.AsUInt64();
```

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128_1.cs]: oops 😐 thanks

**Comment by EgorBo** [comment]
@EgorBot -amd -arm64 -profiler
```cs
using System.IO.Hashing;
using BenchmarkDotNet.Attributes;

public class Bench
{
    static readonly byte[] Data = GetData();

    private static byte[] GetData()
    {
        byte[] bytes = new byte[1000000];
        bytes.AsSpan().Fill(42);
        return bytes;
    }

    [Benchmark]
    public byte[] BenchXxHash128()
    {
        XxHash128 hash = new();
        hash.Append(Data);
        return hash.GetHashAndReset();
    }
}
```

**Comment by EgorBo** [comment]
@EgorBot -amd --envvars DOTNET_EnableAVX:0
```cs
using System.IO.Hashing;
using BenchmarkDotNet.Attributes;

public class Bench
{
    static readonly byte[] Data = GetData();

    private static byte[] GetData()
    {
        byte[] bytes = new byte[1000000];
        bytes.AsSpan().Fill(42);
        return bytes;
    }

    [Benchmark]
    public byte[] BenchXxHash128()
    {
        XxHash128 hash = new();
        hash.Append(Data);
        return hash.GetHashAndReset();
    }
}
```

**Comment by EgorBo** [comment]
@EgorBot -amd -arm64 -profiler
```cs
using System.IO.Hashing;
using BenchmarkDotNet.Attributes;

public class Bench
{
    static readonly byte[] Data = GetData();

    private static byte[] GetData()
    {
        byte[] bytes = new byte[1000000];
        bytes.AsSpan().Fill(42);
        return bytes;
    }

    [Benchmark]
    public byte[] BenchXxHash128()
    {
        XxHash128 hash = new();
        hash.Append(Data);
        return hash.GetHashAndReset();
    }
}
```

**Comment by EgorBo** [comment]
@EgorBot -arm64 -profiler
```cs
using System.IO.Hashing;
using BenchmarkDotNet.Attributes;

public class Bench
{
    static readonly byte[] Data = GetData();

    private static byte[] GetData()
    {
        byte[] bytes = new byte[1000000];
        bytes.AsSpan().Fill(42);
        return bytes;
    }

    [Benchmark]
    public byte[] BenchXxHash128()
    {
        XxHash128 hash = new();
        hash.Append(Data);
        return hash.GetHashAndReset();
    }
}
```

**Comment by EgorBo** [comment]
@EgorBot -amd -intel -arm64 -profiler
```cs
using System.IO.Hashing;
using BenchmarkDotNet.Attributes;

public class Bench
{
    static readonly byte[] Data = new byte[1000000];

    [Benchmark]
    public byte[] BenchXxHash128()
    {
        XxHash128 hash = new();
        hash.Append(Data);
        return hash.GetHashAndReset();
    }
}
```

**Comment by EgorBo** [comment]
@EgorBot -amd -intel -arm64 -profiler --envvars DOTNET_PreferredVectorBitWidth:128
```cs
using System.IO.Hashing;
using BenchmarkDotNet.Attributes;

public class Bench
{
    static readonly byte[] Data = new byte[1000000];

    [Benchmark]
    public byte[] BenchXxHash128()
    {
        XxHash128 hash = new();
        hash.Append(Data);
        return hash.GetHashAndReset();
    }
}
```

**Comment by EgorBo** [comment]
@tannergooding PTAL, I'll add arm64 separately, need to test different impls.
I've expanded it in importer similar to existing op_Multiply expansions

File: src/coreclr/jit/gentree.cpp [hunk 2]
```diff
@@ -21592,11 +21592,55 @@ GenTree* Compiler::gtNewSimdBinOpNode(
                     {
                         intrinsic = NI_AVX10v1_MultiplyLow;
                     }
-                    else
+                    else if (compOpportunisticallyDependsOn(InstructionSet_AVX512DQ_VL))
```
Comment (tannergooding) [inline src/coreclr/jit/gentree.cpp:21595]: Now that we're doing a "proper" check here, we want to combine this and the previous path using:
```cpp
bool isV512Supported = false;
if (compIsEvexOpportunisticallySupported(&isV512Supported, InstructionSet_AVX512DQ_VL))
{
    if (simdSize == 64)
    {
        assert(isV512Supported);
        intrinsic = NI_AVX512DQ_MultiplyLow;
    }
    else
    {
        intrinsic = !isV512Supported ? NI_AVX10v1_MultiplyLow : NI_AVX512DQ_VL_MultiplyLow;
    }
}
```

The reasoning being that `Avx10v1` and `Avx512DQ_VL` are part of the same synthetic `EVEX` baseline and we want the capability of eventually tracking that this is an either/or scenario for R2R/AOT scenarios.

**Review by tannergooding — APPROVED** [review]

File: src/coreclr/jit/hwintrinsicxarch.cpp [hunk 3]
```diff
@@ -2706,17 +2706,25 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
 
             if (varTypeIsLong(simdBaseType))
             {
-                if (simdSize != 64 && !canUseEvexEncoding())
+                if (TARGET_POINTER_SIZE == 4)
                 {
-                    // TODO-XARCH-CQ: We should support long/ulong multiplication
+                    // TODO-XARCH-CQ: 32bit support
```
Comment (tannergooding) [inline src/coreclr/jit/hwintrinsicxarch.cpp:2711]: What's blocking 32-bit support? It doesn't look like we're using any `_X64` intrinsics in the fallback logic?

Comment (EgorBo) [reply src/coreclr/jit/hwintrinsicxarch.cpp:2711]: Not sure to be honest, that check was pre-existing, I only changed comment

File: src/coreclr/jit/gentree.cpp [hunk 4]
```diff
@@ -21592,11 +21592,55 @@ GenTree* Compiler::gtNewSimdBinOpNode(
...
+                        GenTreeHWIntrinsic* tmp2 =
+                            gtNewSimdHWIntrinsicNode(type, op1Dup, tmp1,
+                                                     is256 ? NI_AVX2_MultiplyLow : NI_SSE41_MultiplyLow,
+                                                     CORINFO_TYPE_UINT, simdSize);
+
+                        // Vector256<int> tmp3 = Avx2.HorizontalAdd(tmp2.AsInt32(), Vector256<int>.Zero);
+                        GenTreeHWIntrinsic* tmp3 =
+                            gtNewSimdHWIntrinsicNode(type, tmp2, gtNewZeroConNode(type),
+                                                     is256 ? NI_AVX2_HorizontalAdd : NI_SSSE3_HorizontalAdd,
+                                                     CORINFO_TYPE_UINT, simdSize);
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:21631]: I tried to benchmark different implementations for it and they all were equaly fast e.g. https://github.com/dotnet/runtime/pull/99871#issuecomment-2180529339

File: src/coreclr/jit/gentree.cpp (hunk 2 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp]: Addressed

**Pull request merged**